After talking about the model catalog today, I got curious about the DOIs that have been minted in the USGS ID space that claim to be models. This notebook takes a look. A quick look at the DataCite works API shows that there are quite a few DOIs minted, with the majority coming from work I did to register all those GAP species habitat models. But there seem to be some other interesting things in there. I put in a process to pull all those together by paginating through the API results but then stash them in a Feather file in case for further reference.

In [1]:
import requests
import pandas as pd

recordset = list()

In [2]:
# Function to create lookup URL for USGS DOIs where resource type is model
def get_url(page_number):
    return f"https://api.datacite.org/works?data-center-id=usgs.prod&resource-type-id=model&page[number]={page_number}"


In [3]:
%%time
# Get the first page of results and set up a structure to contain everything
page_number = 1
result = requests.get(get_url(page_number)).json()

CPU times: user 26.7 ms, sys: 7.96 ms, total: 34.6 ms
Wall time: 1.69 s


In [4]:
%%time
# Loop through and build out the full result set (the lazy way)
while len(result["data"]) > 0:
    recordset.extend(result["data"])
    page_number += 1
    result = requests.get(get_url(page_number)).json()

CPU times: user 1.33 s, sys: 104 ms, total: 1.44 s
Wall time: 5min 57s


If we ran the process to get the latest set of model type DOIs, then we build a dataframe from the attributes structure in each record. If we didn't run it, then we pull in the dataframe from the stashed Pickle file.

In [7]:
if len(recordset) == 0:
    df_models = pd.read_pickle("ModelsStash")
else:
    df_models = pd.DataFrame([i["attributes"] for i in recordset])
    df_models.to_pickle("ModelsStash")


In [8]:
# See what the data look like in a dataframe
df_models.head(5)

doi                        identifier  \
0  10.5066/p9hiyvg2  https://doi.org/10.5066/p9hiyvg2   
1  10.5066/f76t0jpb  https://doi.org/10.5066/f76t0jpb   
2  10.5066/f798853m  https://doi.org/10.5066/f798853m   
3  10.5066/f76w98jb  https://doi.org/10.5066/f76w98jb   
4  10.5066/f7h99392  https://doi.org/10.5066/f7h99392   

                                                 url  \
0  https://water.usgs.gov/GIS/metadata/usgswrd/XM...   
1  https://ca.water.usgs.gov/projects/reg_hydro/b...   
2  http://water.usgs.gov/GIS/metadata/usgswrd/XML...   
3  https://water.usgs.gov/GIS/metadata/usgswrd/XM...   
4  http://water.usgs.gov/GIS/metadata/usgswrd/XML...   

                                              author  \
0       [{'given': 'Tracie R', 'family': 'Jackson'}]   
1  [{'given': 'Lorraine E.', 'family': 'Flint'}, ...   
2  [{'given': 'J Hal', 'family': 'Davis'}, {'give...   
3  [{'given': 'Alex R', 'family': 'Fiore'}, {'giv...   
4  [{'given': 'Stephen J', 'family': 'Cauller'}, ...   

                                               title         container-title  \
0  MODFLOW-2005 and PEST models used to character...  U.S. Geological Survey   
1  California Basin Characterization Model: A Dat...  U.S. Geological Survey   
2  MODFLOW 2000 and MT3DMS models of potentiometr...  U.S. Geological Survey   
3  MODFLOW-2005 model used to evaluate the potent...  U.S. Geological Survey   
4  MODFLOW2005 model used to simulate the effects...  U.S. Geological Survey   

                                         description resource-type-subtype  \
0                                               None                 Model   
1  The Basin Characterization Model (BCM), can tr...                 Model   
2  This model is a preliminary characterization o...                 Model   
3  A three-dimensional groundwater flow model, MO...                 Model   
4  A three-dimensional groundwater flow model was...                 Model   

  data-center-id member-id  ... view-count views-over-time download-count  \
0      usgs.prod      usgs  ...          0              []              0   
1      usgs.prod      usgs  ...          0              []              0   
2      usgs.prod      usgs  ...          0              []              0   
3      usgs.prod      usgs  ...          0              []              0   
4      usgs.prod      usgs  ...          0              []              0   

  downloads-over-time published                registered  checked  \
0                  []      2020  2020-03-05T18:30:50.000Z     None   
1                  []      2014  2014-09-17T00:57:18.000Z     None   
2                  []      2016  2016-05-05T18:15:15.000Z     None   
3                  []      2018  2018-03-13T14:32:08.000Z     None   
4                  []      2016  2016-07-26T16:40:24.000Z     None   

                    updated  media  \
0  2020-03-05T18:30:50.000Z     []   
1  2020-03-02T21:34:34.000Z     []   
2  2020-03-02T21:18:55.000Z     []   
3  2020-03-02T21:18:05.000Z     []   
4  2020-03-02T21:16:07.000Z     []   

                                                 xml  
0  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
1  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
2  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
3  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  
4  PD94bWwgdmVyc2lvbj0iMS4wIiBlbmNvZGluZz0iVVRGLT...  

[5 rows x 28 columns]

In [9]:
# Set the option to be able to look at everything in columns for convenience
pd.set_option('display.max_colwidth', None)

One of the things I was interested in is where these DOIs actually point in terms of their dereferencing URLs. To examine this, I create a new column with just the domain part of the URL so that I can group and look at things.

In [10]:
df_models['url_domain'] = df_models.apply(lambda row: row.url.split("/")[2], axis = 1)

In [11]:
df_models[["doi","url_domain"]].groupby('url_domain').count()

doi
url_domain                        
alaska.usgs.gov                  1
axiomdatascience.com             1
ca.water.usgs.gov                1
coastal.er.usgs.gov              1
code.usgs.gov                    2
dx.doi.org                       1
github.com                       1
my.usgs.gov                      3
nrtwq.usgs.gov                   3
regclim.coas.oregonstate.edu     1
usgs.gov                         2
walrus.wr.usgs.gov               1
water.usgs.gov                  93
www.mbr-pwrc.usgs.gov            2
www.sciencebase.gov           1761
www.usgs.gov                     2

That's kind of an interesting spread. Almost all of the 1761 "models" in ScienceBase are the GAP habitat maps, which are a model output of sorts from a habitat affinity-based species distribution modeling method that involves human intervention. But it looks like we have some other interesting things to look at and think about in terms of what they actual represent for a future model catalog. In the sections below, I output particular records, take a look to see what's on the other end, and provide some notes.

In [21]:
# Helper function to filter the dataframe for a particular URL domain
def get_domain_items(domain):
    return df_models.loc[df_models['url_domain'] == domain][["doi","url","title","description"]]

In [18]:
get_domain_items("usgs.gov")

doi               url  title                 description
1874  10.5066/f78050nh   http://usgs.gov  Title                  test111222
1875  10.5066/f7cr5rd0  https://usgs.gov  Title  description of the dataset

Well, those are disappointing. Looks like someone needs to do some cleanup in what actually got registered out with DataCite, if that's even possible.

In [19]:
get_domain_items("axiomdatascience.com")

doi  \
113  10.5066/f72b8w3t   

                                                            url  \
113  http://axiomdatascience.com/maps/integrated/?portal_id=51#   

                                                                                                                           title  \
113  Wave and Wind projections for United States Coasts; Mainland, Pacific Islands, and United States-Affiliated Pacific Islands   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               

This one from a USGS partner in Alaska that I know pretty well looks interesting. The DOI de-references to a web application run by Axiom Data Science that provides an interactive system to visualize and download the cached results (model output data) for a wave height and wind speed model that looks at both historic conditions and projected futures under several scenarios for coastal zones. The abstract and the web site list other interesting parts of the picture, including a USGS Open File Report describing the methods and what is possibly an "off the books" web page that purportedly accompanied a data release but is the kind of thing that might otherwise have ended up in a USGS Data Series or something similar with a journal. This is a pretty interesting case that has just about all the components for a model catalog except that I couldn't find the source code with a cursory look at least.

In [20]:
get_domain_items("code.usgs.gov")

doi  \
344  10.5066/p9nquaow   
481  10.5066/p9nqnh41   

                                                               url  \
344                       https://code.usgs.gov/coawstmodel/COAWST   
481  https://code.usgs.gov/ecosystems/FRESC/sagebrush_hurdle_model   

                           title  \
344  COAWST Modeling System v3.4   
481       sagebrush_hurdle_model   

                                                          description  
344  Coupled ocean atmosphere wave sediment transport modeling system  
481                                                              None

I know a little bit about the COAWST model coupling system as I looked at is as an example for EarthMAP. The DOI reference here points to a fine enough source repository, but it does present some challenges in terms of interfacing toward the model catalog idea. You really have to dig into the user manual in a Word document to get at deeper level information about the model, which means writing code to crawl and put pieces together here would be pretty challenging.

I'd heard about the Sagebrush Hurdle Model but hadn't looked at the source code before. This is an interesting case where there is some useful metadata embedded in the README, namely a reference to a journal article that model was used in. The repo also has a CSV file that contains the input data for the model, which probably happens a fair bit depending on the size of the necessary data. It's not a bad practice, in some ways, to stash input data with the model. However, in this case, there's really not any real documentation in or with the dataset (species occurrence points and basic environmental condition) to understand where it came from without some sleuthing. So, it gets points for transparency, but some demerits for re-usability.

In [22]:
get_domain_items("github.com")

doi                                                 url  \
484  10.5066/p9yjbmbq  https://github.com/rerickson-usgs/eDNA_field_study   

                                                                  title  \
484  Code to assist the USFWS with eDNA field sampling designs for eDNA   

    description  
484        None

This is an interesting case of a minted DOI pointing to a provisional code repo sitting in a personal space in GitHub. I guess that's okay under policy, but it opens a few questions. I would tend to not mint a DOI for something that was not going to eventually be production code. Richie Erickson does provide a nice disclaimer up front in the README indicating that he doesn't expect someone to build on it. There's a title reference to an article that's now [published](https://www.ncbi.nlm.nih.gov/pubmed/31188494), and some guidance about following up to put those links into some type of code metadata that would support the reference would be a good thing to have.

In [23]:
get_domain_items("my.usgs.gov")

doi  \
370   10.5066/p9t9j3ju   
1849  10.5066/f7dj5d4h   
1852  10.5066/f7416v7z   

                                                                                       url  \
370     https://my.usgs.gov/bitbucket/users/rerickson_usgs.gov/repos/networknodeipm/browse   
1849  https://my.usgs.gov/bitbucket/projects/UMESC/repos/migratorypathwaysourcesink/browse   
1852        https://my.usgs.gov/bitbucket/projects/UMESC/repos/migratorypathwaysourcesink/   

                                             title  \
370   Spatially explicit integral projection model   
1849                                carpIPM R code   
1852                 Migratory Pathway Source Sink   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               description  
370                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   None  
1849  Fish grow continuously. Integral projection models (IPM) model size continuously. We developed an IPM for fish. We specifically developed the model for grass carp, however, the model could also be applied to species with similar life-histories such as other carp species. We included YY-males within the model because releasing YY-males has been proposed a control method for the species. YY-males are fish that have 2 male chromosomes compared to a XY-male. When YY-males mate, they only produce male (XY) offspring. This decreases the female proportion of the population and can, in theory, eradicate local populations by biasing the sex-ratio. We created our model as an R package and the model has been documented by both a TRACE Document and peer-reviewed manuscript.  
1852                                                                                                                                                                                                                                                                                                                                                                                                       This is the R code to support the manuscript ``Defining and classifying migratory habitats as sources and sinks: the migratory pathway approach''.\r\nThe are three files besides this file (the README file):\r\nA source file of function: functionsUploaded2016.R\r\nA script file that was used for the manuscript and is an R Markdown file: SubmittedCode.Rmd; and\r\nand a LICENSE file.

These three cases are pointing to the Atlassian BitBucket instance that we've now decomissioned on myUSGS. Someone probably should follow up to see if these repos got moved somewhere else and then change the dereferencing URL in the DOI system. The lack of a description in the DataCite metadata here (and elsewhere) should also be a flag.

In [24]:
get_domain_items("regclim.coas.oregonstate.edu")

doi  \
283  10.5066/f7r78dhk   

                                                            url  \
283  http://regclim.coas.oregonstate.edu/data-access/index.html   

                                                             title description  
283  Dynamically downscaled climate simulations over North America        None

This is a web site that provides access to data distribution services for NetCDF files from a climate model downscaling process. It's essentially model output data or perhaps more precisely, model derivative data. This process has been done in a number of different places using a variety of downscaling methods appropriate to different uses. The most notable framework in USGS for distributing these data with some value-added services for statistical summarization is through the GeoDataPortal. In this case, the DOI might actually be miscategorized as a model but it's related to modeling, indicating that some guidance and parameters are probably needed in USGS policy.

In [25]:
get_domain_items("dx.doi.org")

doi                                 url  \
1853  10.5066/f7639nn9  http://dx.doi.org/10.5066/F7639NN9   

                                            title  \
1853  Capture Map Bias Hypothetical Model Archive   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         description  
1853  A MODFLOW model of a hypothetical stream-aquifer system is presented for the evaluation and characterization of capture map bias.  The hypothetical model is a single-layer model constructed with 30 rows and 100 columns.  The MODFLOW model includes a stream (represented with the MODFLOW SFR or CHD Packages), mountain block recharge (represented with the MODFLOW RCH Package), and evapotransipration (represented with the MODFLOW EVT Package). The hypothetical model is used to create capture maps and capture difference maps.  Map regions with large capture and depletion fraction differences are evaluated with new methods to compute capture map bias.  The hypothetical stream-aquifer system model is used for sensitivity analyses to characterize capture map bias.

This is a curious one that dereferences to another DOI that is not active at the moment. Something weird happened with this. It seems like we should have some safeguards in our toolset to keep this kind of thing from happening.

The description does present an interesting class of model "stuff" that will end up in our catalog somewhere - inputs and outputs from hypothetical model runs made to critically examine aspects of model performance or provide some level of calibration. These are important artifacts in a modeling lifecycle that we should come up with a way to handle, but they should be somehow distinct from other types of model "packages."

In [30]:
# Provide an index on a key substring identifying GAP habitat maps and then show the rest
sciencebase_items = get_domain_items("www.sciencebase.gov")
sciencebase_items["hab_map"] = sciencebase_items["title"].str.find("_2001v1",2)
sciencebase_items.loc[sciencebase_items["hab_map"] < 0]

doi  \
7     10.5066/p9qrd7a3   
9     10.5066/p90jy18d   
10    10.5066/p94owfsm   
16    10.5066/p995smvw   
30    10.5066/f7nk3c59   
38    10.5066/f7kd1x68   
39    10.5066/f7ft8k8k   
40    10.5066/f7qn662b   
41    10.5066/f76972tt   
42    10.5066/f7zc8239   
43    10.5066/f7348jmn   
44    10.5066/f7vt1r95   
45    10.5066/f7sx6bz1   
46    10.5066/f7k35sw0   
47    10.5066/f74b2zfg   
103   10.5066/f7xw4hrj   
104   10.5066/f7t43s20   
105   10.5066/f7n29v2v   
114   10.5066/f7g73cp1   
116   10.5066/f7n58k8w   
117   10.5066/f7t151q4   
168   10.5066/p9843jft   
193   10.5066/p9gt00gb   
235   10.5066/p9k9m1rq   
284   10.5066/f7tx3dmj   
287   10.5066/p93zdx4w   
327   10.5066/p956df36   
342   10.5066/p9k9canl   
414   10.5066/p9ygfmnc   
478   10.5066/p98253jv   
480   10.5066/p9vnzi1w   
482   10.5066/p9g4mpvq   
486   10.5066/p9w22dpc   
493   10.5066/p96xlglp   
498   10.5066/p9yq6h1r   
500   10.5066/p961r9pj   
501   10.5066/p9fbi4u1   
502   10.5066/p9sfuh2k   
503   10.5066/p90swvfu   
1579  10.5066/f7f76b0g   
1838  10.5066/p9qvaxcm   
1839  10.5066/p935y7no   
1873  10.5066/f7wd3xm6   

                                                                    url  \
7     https://www.sciencebase.gov/catalog/item/5c5ddfa2e4b0fe48cb32e717   
9     https://www.sciencebase.gov/catalog/item/5e594162e4b01d50924ab17d   
10    https://www.sciencebase.gov/catalog/item/5e593589e4b01d50924a99ed   
16    https://www.sciencebase.gov/catalog/item/5e3af9b8e4b0edb47bddac83   
30    https://www.sciencebase.gov/catalog/item/57db0908e4b090824ffc3324   
38    https://www.sciencebase.gov/catalog/item/5df12287e4b02caea0f635ca   
39    https://www.sciencebase.gov/catalog/item/5df11bc2e4b02caea0f5f70c   
40    https://www.sciencebase.gov/catalog/item/5df11b10e4b02caea0f5f147   
41    https://www.sciencebase.gov/catalog/item/5df124b1e4b02caea0f6479a   
42    https://www.sciencebase.gov/catalog/item/5df121a8e4b02caea0f62f2d   
43    https://www.sciencebase.gov/catalog/item/5df11ff9e4b02caea0f61f80   
44    https://www.sciencebase.gov/catalog/item/5df11ee8e4b02caea0f61469   
45    https://www.sciencebase.gov/catalog/item/5df1242de4b02caea0f64299   
46    https://www.sciencebase.gov/catalog/item/5df12346e4b02caea0f63bc5   
47    https://www.sciencebase.gov/catalog/item/5df11938e4b02caea0f5e061   
103   https://www.sciencebase.gov/catalog/item/5a6731bae4b06e28e9c55302   
104   https://www.sciencebase.gov/catalog/item/59c452c3e4b017cf313bea5c   
105   https://www.sciencebase.gov/catalog/item/57b12627e4b0fc09fab0ce4f   
114   https://www.sciencebase.gov/catalog/item/5919fd6be4b0a7fdb43b3433   
116   https://www.sciencebase.gov/catalog/item/5994dc46e4b0fe2b9fe915eb   
117   https://www.sciencebase.gov/catalog/item/57f1d4f3e4b0bc0bebfee139   
168   https://www.sciencebase.gov/catalog/item/5b7da9f0e4b045b1dc7beb81   
193   https://www.sciencebase.gov/catalog/item/5d36fd3fe4b01d82ce8a6cd9   
235   https://www.sciencebase.gov/catalog/item/5d964044e4b0c4f70d110f18   
284   https://www.sciencebase.gov/catalog/item/5b5b609ce4b006a11f63f85f   
287   https://www.sciencebase.gov/catalog/item/5d84f914e4b0c4f70d083b86   
327   https://www.sciencebase.gov/catalog/item/5d755283e4b0c4f70d01ad76   
342   https://www.sciencebase.gov/catalog/item/5b6da628e4b0f5d57875f7ea   
414   https://www.sciencebase.gov/catalog/item/5c73ef7ee4b0fe48cb47e513   
478   https://www.sciencebase.gov/catalog/item/5d2337abe4b0941bde64f18b   
480   https://www.sciencebase.gov/catalog/item/5d1f7535e4b0941bde64dbf5   
482   https://www.sciencebase.gov/catalog/item/5d36fe14e4b01d82ce8a6cf0   
486   https://www.sciencebase.gov/catalog/item/5ca6543de4b0c3b0064c26f2   
493   https://www.sciencebase.gov/catalog/item/5c546cffe4b0708288fec2a1   
498   https://www.sciencebase.gov/catalog/item/5bd84f8fe4b0b3fc5ce9d867   
500   https://www.sciencebase.gov/catalog/item/5b6ae1bfe4b006a11f7792b7   
501   https://www.sciencebase.gov/catalog/item/5b966eabe4b0702d0e8268cb   
502   https://www.scienc

I already know about the GAP habitat maps, but I was curious about what other "models" are distributed via ScienceBase. It looks like some cool stuff online. A bunch of missing descriptions is problematic and something that should be easily fixed in the DOI system. Unfortunately, ScienceBase seems to be taking some much needed time off this evening, so I can't pull up any of these. Descriptions look like model outputs, and I would guess that the ScienceBase Items might actually be classed as data releases (or at least they could have been). The nice thing about using ScienceBase as the backend landing system for these DOIs would be the reasonably simple structured metadata capability. I'd be looking for things like links to source code, links to associated publications, and links or relationships to input data that might have been structured into the items, making the production of a linked catalog more feasible.